### PR TITLE
Move redirect handling entirely to PoolManager

### PR DESCRIPTION
**Pull request by Lukasa**
Redirect handling code was duplicated in urllib3, being present in two places: in the ConnectionPool, and the PoolManager. This centralises all redirect handling code into the PoolManager, where it needs to be to handle cross-host redirects. ConnectionPool objects are no longer capable of doing redirects.

After this I'll provide a PR that moves all retry handling up to PoolManager as well, at which point ConnectionPool will have been greatly simplified.

### CODE REVIEW

**Comment by haikuginger** [comment]
Will review in the next couple days.

**Comment by sigmavirus24** [comment]
Am I understanding correctly that this loses the redirect history since we put the onus of managing redirects on the Retry object?

**Comment by Lukasa** [comment]
I don't think so, the tests for RedirectHistory continue to pass.

**Comment by Lukasa** [comment]
Ok, tests are green, this is good to review.

File: urllib3/connectionpool.py [hunk 1]
```diff
@@ -533,11 +521,7 @@ def urlopen(self, method, url, body=None, headers=None, retries=None,
             headers = self.headers
 
         if not isinstance(retries, Retry):
-            retries = Retry.from_int(retries, redirect=redirect, default=self.retries)
-
-        # Check host
-        if assert_same_host and not self.is_same_host(url):
-            raise HostChangedError(self, url, retries)
+            retries = Retry.from_int(retries, default=self.retries)
```
Comment (haikuginger) [inline urllib3/connectionpool.py]: Perhaps an explicit `redirect=False`?

File: urllib3/poolmanager.py [hunk 2]
```diff
@@ -301,8 +302,12 @@ def urlopen(self, method, url, redirect=True, **kw):
         u = parse_url(url)
         conn = self.connection_from_host(u.host, port=u.port, scheme=u.scheme)
 
-        kw['assert_same_host'] = False
-        kw['redirect'] = False
+        # Rewind body position, if needed. Record current position
```
Comment (haikuginger) [inline urllib3/poolmanager.py]: ^update docstring, please.

Comment (Lukasa) [inline urllib3/poolmanager.py]: What docstring update is needed? I feel like I don't understand. 😕

Comment (haikuginger) [inline urllib3/poolmanager.py]: The PoolManager.request method docstring says something about "custom cross-host redirect logic"; this is now the only redirect logic.

Comment (Lukasa) [inline urllib3/poolmanager.py]: Aha, I see. =) Yeah, I can fix that.

**Comment by Lukasa** [comment]
Nits addressed.

**Pull request merged**